streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -3616,33 +3616,33 @@ object ZStream extends ZStreamPlatformSpecificConstructors {
   def fromIterableM[R, E, O](iterable: ZIO[R, E, Iterable[O]]): ZStream[R, E, O] =
     fromEffect(iterable).mapConcat(identity)
 
-  def fromIterator[A](iterator: => Iterator[A]): ZStream[Any, Throwable, A] = {
-    object StreamEnd extends Throwable
-
-    ZStream.fromEffect(Task(iterator) <*> ZIO.runtime[Any]).flatMap { case (it, rt) =>
-      ZStream.repeatEffectOption {
-        Task {
-          val hasNext: Boolean =
-            try it.hasNext
-            catch {
-              case e: Throwable if !rt.platform.fatal(e) =>
-                throw e
-            }
-
-          if (hasNext) {
-            try it.next()
-            catch {
-              case e: Throwable if !rt.platform.fatal(e) =>
-                throw e
-            }
-          } else throw StreamEnd
-        }.mapError {
-          case StreamEnd => None
-          case e         => Some(e)
-        }
-      }
+  /**
+   * Creates a stream from an iterator that may throw exceptions.
+   */
+  def fromIterator[A](iterator: => Iterator[A], maxChunkSize: Int = 1): ZStream[Any, Throwable, A] =
+    ZStream {
+      ZManaged
+        .effect(iterator)
+        .fold(
+          Pull.fail,
+          iterator =>
+            ZIO.effect {
```

Comment:
Since you know the chunk size, it's probably more efficient to pre-allocate an array, fill it, convert to `Chunk`, and `take` the number of elements actually read (which is `<=` the chunk size).

Reply:
Hinting the chunk builder with the desired size will have the same effect and avoid dealing with the array directly

Reply:
So the one watch out here is if the user uses an extremely large `maxChunkSize`. For example, if I say "I'm not sure how many elements I have, but just read them all as one Chunk" and do `fromIterator(iterator, Int.MaxValue)` I am going to blow up with an `OutOfMemoryError` because it is going to try to actually allocate an `Array` with `Int.MaxValue` size.

Reply:
My POV on this is that if it says "chunk" on the label, it's eager and must fit in memory. So I kinda think it's a fair tradeoff.

Reply:
I think that is right but just to push on this that means that `ZStream.fromIterator(Iterator(1, 2, 3, 4, 5), Int.MaxValue)` will blow up. I think definitely if the `Iterator` is actually too large to fit in memory it has to fail, but failing when the `Iterator` is small is a little counterintuitive.

Reply:
True. Maybe we could call it a `chunkHint` and that way we would be free to limit it to some reasonable value (our default chunk size, e.g.).